Paper: Spatial Microsimulation and Activity Allocation in Python: An Update on the Likeness Toolkit (#834)
scipy-conference/scipy_proceedings: papers/james_gaboardi/paper.tex

\subsection{Introduction}\label{introduction}

> Agent-based models linking individual behavior to population dynamics are essential for understanding human security and social equity issues within human systems \cite{germann2022assessing, ozik2021population, urbanpop-AG-2023}. A current challenge for these models is to more directly account for population heterogeneity \cite{BRaVE_DOE_2022}, which relates to factors including socio-demographic characteristics, patterns of life, and critical infrastructure. The UrbanPop spatial microsimulation framework developed by Oak Ridge National Laboratory (ORNL) presents a means to integrate these factors by combining attribute-rich and customizable synthetic populations generated from the American Community Survey (ACS) and its Public-Use Microdata Sample (PUMS) with real-world data on housing, transportation, and points of interest (POIs) \cite{urbanpop-AG-2023}. Central to UrbanPop's capabilities is Likeness, a Python toolkit that combines population synthesis, spatial network modeling, and activity allocation \cite{likeness-scipy-paper-2022, likeness-scipy-poster-2022}. Likeness supports the creation of more dynamic agent-based models that can be applied to topics ranging from epidemiology to environmental hazards. 

Reviewer: The introduction of the paper provides a clear overview of the research topic and highlights the significance of agent-based models in understanding human security and social equity issues. Here are some suggestions to improve the different aspects of the introduction:

- The introduction is well-written and effectively communicates the main points. However, you could enhance the clarity and flow by breaking down the long sentences into shorter ones. Additionally, consider using more specific and concise language to avoid unnecessary repetition.
- Lastly, in the first sentence, you could add a bit more detail about what agent-based models are and why they are important. For example, you could say something like "Agent-based models are computer simulations that allow us to model the behavior of individual agents and how their interactions can give rise to complex system-level behavior."

Reviewer 2: - [x] @jvtcl to address this.

> Agent-based models linking individual behavior to population dynamics are essential for understanding human security and social equity issues within human systems \cite{germann2022assessing, ozik2021population, urbanpop-AG-2023}. A current challenge for these models is to more directly account for population heterogeneity \cite{BRaVE_DOE_2022}, which relates to factors including socio-demographic characteristics, patterns of life, and critical infrastructure. The UrbanPop spatial microsimulation framework developed by Oak Ridge National Laboratory (ORNL) presents a means to integrate these factors by combining attribute-rich and customizable synthetic populations generated from the American Community Survey (ACS) and its Public-Use Microdata Sample (PUMS) with real-world data on housing, transportation, and points of interest (POIs) \cite{urbanpop-AG-2023}. Central to UrbanPop's capabilities is Likeness, a Python toolkit that combines population synthesis, spatial network modeling, and activity allocation \cite{likeness-scipy-paper-2022, likeness-scipy-poster-2022}. Likeness supports the creation of more dynamic agent-based models that can be applied to topics ranging from epidemiology to environmental hazards. 

Reviewer 3: A clear definition or equivalent term for 'agent' would be appreciated in your introduction/overview. Perhaps mentioning 'agents' are 'simulated individuals' - could be obvious for the authors but not the reader.

Reviewer 3: A similar case can be made for another keyword - 'population dynamics' - to convey the purpose.

Reviewer 2: - [x] @jvtcl to address this

Reviewer 2: - Generalized "agent-based model" to "synthetic population" (and removed "population synthesis")
- Changed "human dynamics" -> "population dynamics" to match the text
- Added "microsimulation"

> Foundational examples produced for Likeness have explored travel routing from home to anchor activities (e.g., work, school), POI occupancy characteristics \cite{likeness-scipy-paper-2022, likeness-scipy-poster-2022}, as well as routing incidental travel from anchor to non-anchor activities (e.g., social, errands, health) \cite{likeness_aag_2023}. This paper builds from these foundational examples to

Reviewer: Consider highlighting particular societal or scientific issues that can be solved using the developed methodologies, in my opinion, to further emphasize the paper's significance. Be sure to emphasize the insights that may be gained from the combination of synthetic populations, real-world data, and activity allocation in a variety of fields, including urban planning, public health, and disaster preparedness. I think this will help readers to find the introduction more interesting, informative, and appealing.

Reviewer 3: I agree with this. The work is technically sound, but will benefit from highlighting the more obvious implications and insights to appeal to the general audience. @jGaboardi you can perhaps attend to this at the end of your abstract, Introduction and/or conclusion.

Reviewer 2: - [x] @jvtcl to address this.

> Foundational examples produced for Likeness have explored travel routing from home to anchor activities (e.g., work, school), POI occupancy characteristics \cite{likeness-scipy-paper-2022, likeness-scipy-poster-2022}, as well as routing incidental travel from anchor to non-anchor activities (e.g., social, errands, health) \cite{likeness_aag_2023}. This paper builds from these foundational examples to

Reviewer 3: I could not find the work associated with likeness_aag_2023. Is this published or currently under review? If so, please indicate in References.

Reviewer 2: - [x] @jGaboardi to address this. I think there's a Zenodo for this? We might have just forgotten to add the BibTeX.

Author: We have a Zenodo DOI for it and it is listed in out references section:

> [7] J. D. Gaboardi and J. V. Tuccillo, “Spatial Microsimulation and ActivityAllocation  for  Examining  COVID-19  Vaccine  Access  Profiles,”  Mar.2023.

It is a conference presentation with designated DOI. However, the BiBteX style was omitting the `url` field. I updated the entry with a `howpublished` field that now shows the proper URL.

@jvtcl Please mark as completed.

> Foundational examples produced for Likeness have explored travel routing from home to anchor activities (e.g., work, school), POI occupancy characteristics \cite{likeness-scipy-paper-2022, likeness-scipy-poster-2022}, as well as routing incidental travel from anchor to non-anchor activities (e.g., social, errands, health) \cite{likeness_aag_2023}. This paper builds from these foundational examples to

Reviewer 3: It may help to have a clear distinction between 'anchor' and 'non-anchor' POIs. For example, is 'home' considered anchor as implied from "anchor to non-anchor activities (e.g. social)..."?

Reviewer 2: @ameyxd We consider home to be separate from anchor activities like school and work. We will clarify by using the terminology "home locations". 

- [x] @jvtcl to address this.

> discuss scaling our methods to larger areas of interest, as well as examines the performance of our activity allocation model relative to real-world POI demographics from digital trace data.

Reviewer 3: Please define or expand on the phrase 'digital trace data'.

Reviewer 2: - [x] @jvtcl to address this.

Reviewer 2: I added a brief clarification that this refers to "anonymized mobile device" data.

\subsection{Expansion and Scaling of Likeness Capabilities}\label{section:likeness-expansion-scaling}

\input figures/ecosystem.tex

> In step with the creation of foundational examples for Likeness, we have developed an integrated workflow that scales the ecosystem (Figure~\ref{fig:ecosystem}) to support microsimulation for any metropolitan statistical area (MSA) in the United States.

Reviewer 3: I do not recommend starting with a preposition here. "Synchronous with" or similar may be better.

Reviewer 2: - [x] @jvtcl to address this.

> Figure~\ref{fig:parse-to-attribution} outlines the procedure for agent generation at the MSA level. As discussed in \cite{likeness-scipy-paper-2022}, the \texttt{livelike.acs.puma} class is the core object for residential population synthesis, which stores all census microdata and geographic (e.g., block group, tract) model constraints to support spatial allocation for a single Public-Use Microdata Area (PUMA) in the United States. Residential synthetic populations for MSAs are generated as collections of PUMAs, which are parsed automatically by combining U.S. Census metropolitan/micropolitan delineation files\footnote{https://www.census.gov/geographies/reference-files/time-series/demo/metro-micro/delineation-files.html} with the Census 2010 PUMA-to-tract relationship file\footnote{https://www.census.gov/programs-surveys/geography/technical-documentation/records-layout/2010-tract-to-puma-record-layout.html}. PUMAs parsed for the target MSA are converted into \texttt{livelike.acs.puma} objects in bulk via \texttt{livelike.multi.make\_pumas()}, which accepts the PUMA FIPS codes in list form, along with the target ACS year, and uses this information to gather the relevant ACS Summary File (SF) constraints for spatial allocation. Spatial allocation is subsequently handled with parallelization using Python's built-in \texttt{multiprocessing} utility, which is supported by both packages Likeness offers for Penalized Maximum-Entropy Dasymetric Modeling (P-MEDM) \cite{nagle2014dasymetric}: \texttt{pymedm} (bleeding-edge, Python-native version, based on \texttt{jaxopt} \cite{jaxopt_implicit_diff}) and \texttt{pmedm\_legacy} (stable bridge to original R/C++ routine\footnote{https://bitbucket.org/nnnagle/pmedmrcpp} via \texttt{rpy2}). As demonstrated in \cite{likeness-scipy-paper-2022} and \cite{likeness-scipy-poster-2022}, the population synthesis routine also collects diagnostics on each P-MEDM solution's degree of conformity to the published 90\% Margins of Error (MOEs) reported in the ACS SF (``MOE Fit Rate''), available in both \texttt{pymedm} and \texttt{pmedm\_legacy}. 

Reviewer: - I think it would be good to provide more details about the validation exercises that conducted. For example, You could discuss the specific metrics that they used to measure the accuracy of the simulations, and they could provide more information about the results of the validation exercises.

Reviewer 2: - [x] @jvtcl to add a citation for CCA [here] (https://github.com/jGaboardi/scipy_proceedings/blob/f49ce084823d019f9c7458534a7211078f99f2c4/papers/james_gaboardi/paper.tex#L68).

@kalyan678 the highlighted text does not seem to match the section that you'd like us to update. Could you please point us to the lines in the text that need revision?

Reviewer: @jvtcl - Sorry for the confusion. I completely misplaced my comment.
I think you made a point in the abstract and on page 5 of the study about conducting validation exercises. These were the lines I was referring to. (We were also
limited to only one month of POI visit data, but we hope
to expand its scope for future validation exercises)

> Figure~\ref{fig:parse-to-attribution} outlines the procedure for agent generation at the MSA level. As discussed in \cite{likeness-scipy-paper-2022}, the \texttt{livelike.acs.puma} class is the core object for residential population synthesis, which stores all census microdata and geographic (e.g., block group, tract) model constraints to support spatial allocation for a single Public-Use Microdata Area (PUMA) in the United States. Residential synthetic populations for MSAs are generated as collections of PUMAs, which are parsed automatically by combining U.S. Census metropolitan/micropolitan delineation files\footnote{https://www.census.gov/geographies/reference-files/time-series/demo/metro-micro/delineation-files.html} with the Census 2010 PUMA-to-tract relationship file\footnote{https://www.census.gov/programs-surveys/geography/technical-documentation/records-layout/2010-tract-to-puma-record-layout.html}. PUMAs parsed for the target MSA are converted into \texttt{livelike.acs.puma} objects in bulk via \texttt{livelike.multi.make\_pumas()}, which accepts the PUMA FIPS codes in list form, along with the target ACS year, and uses this information to gather the relevant ACS Summary File (SF) constraints for spatial allocation. Spatial allocation is subsequently handled with parallelization using Python's built-in \texttt{multiprocessing} utility, which is supported by both packages Likeness offers for Penalized Maximum-Entropy Dasymetric Modeling (P-MEDM) \cite{nagle2014dasymetric}: \texttt{pymedm} (bleeding-edge, Python-native version, based on \texttt{jaxopt} \cite{jaxopt_implicit_diff}) and \texttt{pmedm\_legacy} (stable bridge to original R/C++ routine\footnote{https://bitbucket.org/nnnagle/pmedmrcpp} via \texttt{rpy2}). As demonstrated in \cite{likeness-scipy-paper-2022} and \cite{likeness-scipy-poster-2022}, the population synthesis routine also collects diagnostics on each P-MEDM solution's degree of conformity to the published 90\% Margins of Error (MOEs) reported in the ACS SF (``MOE Fit Rate''), available in both \texttt{pymedm} and \texttt{pmedm\_legacy}. 

Reviewer 3: FIPS codes are not defined or explained. If relevant, please expand here.

Reviewer 2: - [x] @jvtcl to address this.

Likeness generates agents for microsimulation in a way that provides realistic home (origin) locations from which to allocate essential activities on transportation networks. Our initial approach, based on census block-level housing density \cite{likeness-scipy-paper-2022, likeness-scipy-poster-2022}, is now implemented in \texttt{livelike} as a housing universe generation procedure. Additionally, we are actively developing a method (demonstrated in \nameref{section:LCFL}) that enhances this capability by matching synthesized households to residential locations. These matches are performed via building footprint data by conflating housing type labels between synthetic households (linked from PUMS) and structures including single-family and multi-family dwellings. 

\input figures/parse-to-attribution

Agent generation at residential locations via \texttt{livelike} provides origin points for simulating travel to essential activities. 

> The next stage in the workflow employs network analysis to model the cost of travel to these activities \cite{OpenStreetMap, osmnx_CEUS_2017, foti_generalized_2012} and allocate agents to POIs accordingly based on mathematical programming routines \cite{mitchell_pulp_2011, santos_mixed_2020, lougee_coin_2003, forrest_coinorcbc_2023}. In our first iteration of Likeness, both these tasks were accomplished within \texttt{actlike}. However, it was decided that the network modeling piece was specialized enough to be split from the \texttt{actlike} package, which led to the creation of \texttt{movelike}. With a push for varied modes of network traversal, three new modes of travel can now be modeled: walking, biking, public transit. However, modeling travel behavior via public transportation is less straightforward than for driving, biking, and walking networks due to stricter network topology, including factors like connectivity and directionality of routes. We have taken our first foray into modeling more realistic public transit behavior within \texttt{movelike} through the incorporation of the General Transit Feed Specification (GTFS)\footnote{https://gtfs.org/}. GTFS is a data specification that stipulates the required files, along with their structure and format\footnote{https://gtfs.org/schedule/reference/\#dataset-files}, for publishing, ingesting, and utilizing public transit datasets. The GTFS datasets can be obtained via services such as \textit{TransitFeeds}\footnote{https://transitfeeds.com/} and \textit{The Mobility Database Catalogs}\footnote{https://github.com/MobilityData/mobility-database-catalogs}. In our current iteration we utilize GTFS data feeds to implement a pseudo-transit network space by which agents can engage in limited traversal. This is accomplished through a mask of \textit{OpenStreetMap}\footnote{https://www.openstreetmap.org/} (OSM) street segments known to be associated with bus routes. The OSM network is masked by passing a (multi)polygon feature of buffered and unioned bus routes within the study area into \texttt{osmnx} \cite{osmnx_CEUS_2017}. This method demonstrates progression in representing public transit but certainly has room for improvement, which will be discussed in \nameref{section:dev-roadmap}.

Reviewer 3: Please expand on or mention the programming routines/libraries referenced here: PuLP, MILP, COIN-OR, etc.

Reviewer 2: - [x] @jGaboardi to address this.

Author: @jvtcl I expanded on this within a footnote in our internal ShareLeaf.

Author: @jvtcl ~~please mark complete~~ just realized you already did that.

> The next stage in the workflow employs network analysis to model the cost of travel to these activities \cite{OpenStreetMap, osmnx_CEUS_2017, foti_generalized_2012} and allocate agents to POIs accordingly based on mathematical programming routines \cite{mitchell_pulp_2011, santos_mixed_2020, lougee_coin_2003, forrest_coinorcbc_2023}. In our first iteration of Likeness, both these tasks were accomplished within \texttt{actlike}. However, it was decided that the network modeling piece was specialized enough to be split from the \texttt{actlike} package, which led to the creation of \texttt{movelike}. With a push for varied modes of network traversal, three new modes of travel can now be modeled: walking, biking, public transit. However, modeling travel behavior via public transportation is less straightforward than for driving, biking, and walking networks due to stricter network topology, including factors like connectivity and directionality of routes. We have taken our first foray into modeling more realistic public transit behavior within \texttt{movelike} through the incorporation of the General Transit Feed Specification (GTFS)\footnote{https://gtfs.org/}. GTFS is a data specification that stipulates the required files, along with their structure and format\footnote{https://gtfs.org/schedule/reference/\#dataset-files}, for publishing, ingesting, and utilizing public transit datasets. The GTFS datasets can be obtained via services such as \textit{TransitFeeds}\footnote{https://transitfeeds.com/} and \textit{The Mobility Database Catalogs}\footnote{https://github.com/MobilityData/mobility-database-catalogs}. In our current iteration we utilize GTFS data feeds to implement a pseudo-transit network space by which agents can engage in limited traversal. This is accomplished through a mask of \textit{OpenStreetMap}\footnote{https://www.openstreetmap.org/} (OSM) street segments known to be associated with bus routes. The OSM network is masked by passing a (multi)polygon feature of buffered and unioned bus routes within the study area into \texttt{osmnx} \cite{osmnx_CEUS_2017}. This method demonstrates progression in representing public transit but certainly has room for improvement, which will be discussed in \nameref{section:dev-roadmap}.

Reviewer 3: ```suggestion
The next stage in the workflow employs network analysis to model the cost of travel to these activities \cite{OpenStreetMap, osmnx_CEUS_2017, foti_generalized_2012} and allocate agents to POIs accordingly based on mathematical programming routines \cite{mitchell_pulp_2011, santos_mixed_2020, lougee_coin_2003, forrest_coinorcbc_2023}. In our first iteration of Likeness, both these tasks were accomplished within \texttt{actlike}. However, we concluded that the network modeling piece was specialized enough to be split from the \texttt{actlike} package, which led to the creation of \texttt{movelike}. With a push for varied modes of network traversal, three new modes of travel can now be modeled: walking, biking, public transit. However, modeling travel behavior via public transportation is less straightforward than for driving, biking, and walking networks due to stricter network topology, including factors like connectivity and directionality of routes. We have made our foray into modeling more realistic public transit behavior within \texttt{movelike} through the incorporation of the General Transit Feed Specification (GTFS)\footnote{https://gtfs.org/}. GTFS is a data specification that stipulates the required files, along with their structure and format\footnote{https://gtfs.org/schedule/reference/\#dataset-files}, for publishing, ingesting, and utilizing public transit datasets. The GTFS datasets can be obtained via services such as \textit{TransitFeeds}\footnote{https://transitfeeds.com/} and \textit{The Mobility Database Catalogs}\footnote{https://github.com/MobilityData/mobility-database-catalogs}. In our current iteration we utilize GTFS data feeds to implement a pseudo-transit network space by which agents can engage in limited traversal. This is accomplished through a mask of \textit{OpenStreetMap}\footnote{https://www.openstreetmap.org/} (OSM) street segments known to be associated with bus routes. The OSM network is masked by passing a (multi)polygon feature of buffered and unioned bus routes within the study area into \texttt{osmnx} \cite{osmnx_CEUS_2017}. This method demonstrates progression in representing public transit but certainly has room for improvement, which will be discussed in \nameref{section:dev-roadmap}.
```

Reviewer 3: Grammar and phrasing - 'made our foray', etc. 'first foray' is redundant - foray implies first attempt.

Reviewer 2: - [ ] @jGaboardi to address this.

> The next stage in the workflow employs network analysis to model the cost of travel to these activities \cite{OpenStreetMap, osmnx_CEUS_2017, foti_generalized_2012} and allocate agents to POIs accordingly based on mathematical programming routines \cite{mitchell_pulp_2011, santos_mixed_2020, lougee_coin_2003, forrest_coinorcbc_2023}. In our first iteration of Likeness, both these tasks were accomplished within \texttt{actlike}. However, it was decided that the network modeling piece was specialized enough to be split from the \texttt{actlike} package, which led to the creation of \texttt{movelike}. With a push for varied modes of network traversal, three new modes of travel can now be modeled: walking, biking, public transit. However, modeling travel behavior via public transportation is less straightforward than for driving, biking, and walking networks due to stricter network topology, including factors like connectivity and directionality of routes. We have taken our first foray into modeling more realistic public transit behavior within \texttt{movelike} through the incorporation of the General Transit Feed Specification (GTFS)\footnote{https://gtfs.org/}. GTFS is a data specification that stipulates the required files, along with their structure and format\footnote{https://gtfs.org/schedule/reference/\#dataset-files}, for publishing, ingesting, and utilizing public transit datasets. The GTFS datasets can be obtained via services such as \textit{TransitFeeds}\footnote{https://transitfeeds.com/} and \textit{The Mobility Database Catalogs}\footnote{https://github.com/MobilityData/mobility-database-catalogs}. In our current iteration we utilize GTFS data feeds to implement a pseudo-transit network space by which agents can engage in limited traversal. This is accomplished through a mask of \textit{OpenStreetMap}\footnote{https://www.openstreetmap.org/} (OSM) street segments known to be associated with bus routes. The OSM network is masked by passing a (multi)polygon feature of buffered and unioned bus routes within the study area into \texttt{osmnx} \cite{osmnx_CEUS_2017}. This method demonstrates progression in representing public transit but certainly has room for improvement, which will be discussed in \nameref{section:dev-roadmap}.

Reviewer 3: Please cite OpenStreetMap for all figures outside of footnotes.

Reviewer 2: @ameyxd Could you please clarify this point? Do you mean adding attribution to captions for figures that use OSM?

- [ ] @jGaboardi  to address this.

Author: @ameyxd All plots that use OSM are already attributed within the figure themselves. Are you suggesting another layer of attritbution/citation?

Reviewer 3: I noticed that. You can disregard this - all OSM attributions in figures are good.

> Finally -- and at the heart of it all -- expansion and scaling of the Likeness ecosystem led to the development of a new package for common utilities, \texttt{likeness-vitals}, which provides support for monitoring and timing processes, data manipulation, shared spatial functionality, and Census API access.

Reviewer: - The future use of the Likeness toolset has been outlined in clear terms, and I appreciated the thorough citations of pertinent academic works.
- Overall, I believe that we can simplify the language and sentence structure to make the content more accessible to readers. Break down complex sentences into smaller, more concise ones.

Reviewer 2: - [x] @jvtcl to address this.

Reviewer 2: I made a runthrough of the manuscript and broke up sentences where possible.

\subsection{Integrated Demonstration: Leon County, Florida} \label{section:LCFL}

> Following the workflow described in Section~\nameref{section:likeness-expansion-scaling}, we demonstrate the current capabilities of Likeness and validate our activity allocation routine for Leon County, Florida. Leon County, whose primary city is Tallahassee, features a population of just under 300,000 residents, a compact urban footprint, and a diverse array of transportation modes (driving, transit, bike, walking). Our mobility validation exercise is based on grocery store visits from simulated home locations. Grocery stores provide a useful test case because they are both catchments for the general population and points of access to vital services including food and healthcare. We obtained grocery store visits from Foursquare's Research Visits feed, which provides footfall data attributed by demographic cohort (gender by age) for a variety of point of interest (POI) types\footnote{https://location.foursquare.com/places/docs/how-does-places-work}\textsuperscript{,}\footnote{https://location.foursquare.com/visits/docs/research-feed-schema}. 

Reviewer: - Although you provided a reference to TransitFeeds data, I double-checked and discovered that "TransitFeeds data is not regularly updated. Find up-to-date GTFS Schedule data on the the Mobility Database catalogs." I do not see any such highlight in the paper.

Reviewer 2: - [x] @jGaboardi to address this.

Author: @kalyan678 

In that sentence we also provide a comment on The Mobility Database Catalogs and a footnote including a link to the GitHub Repo (should be footnote (8.)). Is there something more you are looking for here?

> The GTFS datasets can be obtained via services such as *TransitFeeds*[^7] and *The Mobility Database Catalogs*[^8].

> [^7]: https://transitfeeds.com/
> [^8]: https://github.com/MobilityData/mobility-database-catalogs

Reviewer: @jGaboardi Thank you for your clarification. I've seen the specifics, and I believe we're in good hands here :)

Author: @jvtcl Please mark as completed.

> Following the workflow described in Section~\nameref{section:likeness-expansion-scaling}, we demonstrate the current capabilities of Likeness and validate our activity allocation routine for Leon County, Florida. Leon County, whose primary city is Tallahassee, features a population of just under 300,000 residents, a compact urban footprint, and a diverse array of transportation modes (driving, transit, bike, walking). Our mobility validation exercise is based on grocery store visits from simulated home locations. Grocery stores provide a useful test case because they are both catchments for the general population and points of access to vital services including food and healthcare. We obtained grocery store visits from Foursquare's Research Visits feed, which provides footfall data attributed by demographic cohort (gender by age) for a variety of point of interest (POI) types\footnote{https://location.foursquare.com/places/docs/how-does-places-work}\textsuperscript{,}\footnote{https://location.foursquare.com/visits/docs/research-feed-schema}. 

Reviewer 3: This is a general comment for the reference to sections: When using \nameref, try to number sections to reduce redundancy and verbosity, if the paper format allows for it (by setting \documentclass[conference] or by changing \setcounter). This goes for references to sections throughout the document.

Reviewer 2: - [ ] @jGaboardi to address this.

Author: @ameyxd We are not sure that the SciPy Proceedings allow for that and we could not find examples of that format in other submissions for this year. Perhaps we can get an editor's opinion on the matter? cc @mepa

Author: We will retain our `\nameref` format that is compatible with the SciPy proceeding template unless specifically guided to do otherwise.

> Following the workflow described in Section~\nameref{section:likeness-expansion-scaling}, we demonstrate the current capabilities of Likeness and validate our activity allocation routine for Leon County, Florida. Leon County, whose primary city is Tallahassee, features a population of just under 300,000 residents, a compact urban footprint, and a diverse array of transportation modes (driving, transit, bike, walking). Our mobility validation exercise is based on grocery store visits from simulated home locations. Grocery stores provide a useful test case because they are both catchments for the general population and points of access to vital services including food and healthcare. We obtained grocery store visits from Foursquare's Research Visits feed, which provides footfall data attributed by demographic cohort (gender by age) for a variety of point of interest (POI) types\footnote{https://location.foursquare.com/places/docs/how-does-places-work}\textsuperscript{,}\footnote{https://location.foursquare.com/visits/docs/research-feed-schema}. 

Reviewer 3: ```suggestion
Following the workflow described in Section~\nameref{section:likeness-expansion-scaling}, we demonstrate the current capabilities of Likeness and validate our activity allocation routine for Leon County, Florida. Leon County, whose primary city is Tallahassee, features a population of just under 300,000 residents, a compact urban footprint, and a diverse array of transportation modes (driving, transit, bike, walking). Our mobility validation exercise is based on grocery store visits from simulated home locations. Grocery stores provide a useful test case because they are both catchments for the general population and points of access to vital services including food and healthcare. We obtained grocery store visits from Foursquare's Research Visits feed, which provides footfall data attributed by demographic cohort (gender by age) for a variety of POI types\footnote{https://location.foursquare.com/places/docs/how-does-places-work}\textsuperscript{,}\footnote{https://location.foursquare.com/visits/docs/research-feed-schema}. 
```

Reviewer 3: POI is defined.

Reviewer 2: - [ ] @jGaboardi to address this.

> We first simulated a single synthetic population for the Tallahassee Core-Based Statistical Area (CBSA), based on constraining variables including technical characteristics linked to sampling universe totals (i.e., population, housing units, households), as well as descriptive factors including demographics, socioeconomic status, housing, mobility, and worker and student characteristics. For the remainder of the analysis, we focused on Leon County alone, removing large outlying areas of the MSA (PUMA \texttt{1206300}). Agents\footnote{Agents less than 20 years old were not include in the Foursquare POI data, thus they are excluded from our analysis.} were generated on the Federal Emergency Management Agency's (FEMA) open USA Structures database \cite{yang2018building, usa_struct_2022}, based on the sampling of synthetic households to single-family residential, multi-family residential, mobile homes, group quarters housing types.

Reviewer: There are a few technical terms and acronyms that could be briefly explained or replaced with simpler language to improve understanding for a broader audience. For example, instead of "PUMA 1206300," you could mention that it refers to a specific geographical area.

Reviewer 3: I disagree, perhaps add geographical area but don’t remove the PUMA. It is descriptive and helps people familiar with the topic quickly identify the region if they want to.

Reviewer 2: - [x] @jvtcl to clarify this.

Reviewer: I apologise for the confusion. It is not my intention to remove the PUMA. The goal is to add the geographical area rather than the code.

Reviewer 2: Added the [PUMA name](https://censusreporter.org/profiles/79500US1206300-apalachee-region-outside-leon-county-puma-fl/) as provided by Census.

> We first simulated a single synthetic population for the Tallahassee Core-Based Statistical Area (CBSA), based on constraining variables including technical characteristics linked to sampling universe totals (i.e., population, housing units, households), as well as descriptive factors including demographics, socioeconomic status, housing, mobility, and worker and student characteristics. For the remainder of the analysis, we focused on Leon County alone, removing large outlying areas of the MSA (PUMA \texttt{1206300}). Agents\footnote{Agents less than 20 years old were not include in the Foursquare POI data, thus they are excluded from our analysis.} were generated on the Federal Emergency Management Agency's (FEMA) open USA Structures database \cite{yang2018building, usa_struct_2022}, based on the sampling of synthetic households to single-family residential, multi-family residential, mobile homes, group quarters housing types.

Reviewer 3: ```suggestion
We first simulated a single synthetic population for the Tallahassee Core-Based Statistical Area (CBSA), based on constraining variables including technical characteristics linked to sampling universe totals (i.e., population, housing units, households), as well as descriptive factors including demographics, socioeconomic status, housing, mobility, and worker and student characteristics. For the remainder of the analysis, we focused on Leon County alone, removing large outlying areas of the MSA (PUMA \texttt{1206300}). Agents\footnote{Agents less than 20 years old were not include in the Foursquare POI data, thus they are excluded from our analysis.} were generated on the Federal Emergency Management Agency's (FEMA) open USA Structures database \cite{yang2018building, usa_struct_2022}, based on the sampling of synthetic households to single-family residential, multi-family residential, mobile homes, and group quarters housing types.
```

Reviewer 2: - [ ] @jGaboardi to address this.

As demonstrated in Table~\ref{tab:agents_counts_travel_mode}, the `drive' network supports the overwhelming majority of travel to grocery POIs in Leon County, followed by walking, public transportation, and bicycle access. 

\input tables/agents_counts_travel_mode

> Figure~\ref{fig:2x2_agents_hex} shows that Leon County's agent population is distributed unevenly relative to assigned travel modes. Because Leon County's infrastructure primarily supports travel by car, agents who drive are distributed closest to the area's general population density. The spatial distribution of agents who travel by walking also tends to follow Leon County's settlement patterns, though in more limited numbers than for those who drive. Agents using public transport, meanwhile, are largely present in and near the center of the county, roughy occupying denser urban areas where StarMetro service is available. Bicyclists are distributed similarly to bus takers, but with several individual clusters associated with smaller outlying towns and settlements. 

Reviewer 3: The readability and analysis of Figure 4 could be improved by moving it to page 3.

Reviewer 2: - [ ] @jGaboardi to address this.

Author: After we address and resolve all text edits, etc. we will attempt to accommodate this suggestion. However, the typesetting is done within LaTeX and we have limited control over the resultant PDF output.

Reviewer 3: You may be able to force positions. It's okay if not. 👍🏽

Author: I spent some time trying to manually place the figures as suggested, but in doing so they both (a) didn't fit properly in the desired space; and (b) completely threw off the typesetting of the other figures & tables. We will leave as is.

* @jvtcl Please mark complete

\input figures/2x2_agents_hex

Grocery store POIs with medium to high visit confidence (at least 30 device visits per month, $n$ = 53) were obtained from Foursquare for Leon County in January 2023. Destination capacities were estimated based on visit counts weighted by representativeness of the demographic cohort within the state's 2010 Census population\footnote{https://location.foursquare.com/visits/docs/foursquare-data-normalization}. Destination capacities were estimated by the daily average (mean) for each POI during the collection month (01/2023).

> After travel mode was assigned, four network cost matrices were calculated from origins (agents' residential locations) to destination POIs in \texttt{movelike}. Agents were then allocated to a single probable destination with the \texttt{actlike.ActivityAllocation} routine, which solves a modified Transportation Problem \cite{tp_miller_gc__2015}, where destination POI capacities are scaled \cite{lovelace_truncate_2013} by the proportion of assigned travel mode for each scenario. All models were run consecutively on two machines for benchmarking purposes. These were:

Reviewer 3: Please elaborate on the original Transportation Problem. Is it based on the GIS-T? The referenced citation is broad.

Reviewer 2: - [x] @jGaboardi to address this.

Author: * I have added several more (and clearer) citation within our text and added some clarifying language.
* Here we are referring to and expanding on the [Hitchcock/Koopmans Transportation Problem](https://en.wikipedia.org/wiki/Transportation_theory_(mathematics)#Hitchcock_problem).
* @jvtcl Please mark as completed.

\begin{itemize}
    \item A personal laptop (macOS) with a 2.3 GHz Quad-core Intel Core i7 processor (32 GB RAM).
    \item A virtual machine (Ubuntu) with a 2.8 GHz 22-core Intel(R) Xeon(R) processor (86 GB RAM).
\end{itemize}
    

> The large disparity in problem size seen in Table~\ref{tab:agents_counts_travel_mode} is even more pronounced in solution runtimes, shown in Table~\ref{tab:allocation_solution_runtime}, with optimal solutions for non-drive models being found in a maximum time of just over 1 minute on both machines and the drive model taking more than 17 and 8 hours to solve on the macOS and Ubuntu machines, respectively. Considering the solution time for the drive scenario, there is clearly a need for a more effective solution technique, which will be further discussed in \nameref{section:dev-roadmap}.

Reviewer 3: In line with the suggestion to rearrange figures, consider showing Table 1 and Table 2 on the same page to highlight clear correlations in their content (high agent count corresponding to high runtime).

Reviewer 2: - [ ] @jGaboardi to address this.

Author: As [commented above](https://github.com/scipy-conference/scipy_proceedings/pull/834#discussion_r1242936515), we will attempt to accommodate this suggestion.

Author: See comment above --> https://github.com/scipy-conference/scipy_proceedings/pull/834#discussion_r1256470229

* @jvtcl Please mark complete

\input tables/allocation_solution_runtime

To enable comparison with the results of \cite{likeness-scipy-paper-2022} and \cite{urbanpop-AG-2023}, which focused on anchor activities (work/school travel), we validated grocery store visits using Canonical Correlation Analysis (CCA) on the between-destination (relative prevalence) and within-destination (compositional) characteristics of each POI by demographic group. Both CCA runs were generated from tabulated counts of trips from the observed and synthetic datasets, their key difference being the method of standardization (column-wise for between-destination, row-wise for within-destination). We used the CCA coefficient of determination ($R^2$) to measure associations between synthetic and observed results. 

> We also investigated local (POI-specific) within-destination correspondence in compositional characteristics based on gender/age cohorts in five-year intervals (until age 70, which is top-coded), in this case based on Spearman rank correlation due to the relatively small number of demographic cohorts ($n$ = 11).

Reviewer 3: Please cite Spearman correlation here.

Reviewer 2: - [x] @jvtcl to address this.

\subsection{Results}\label{section:results}

> \subsubsection{Demographic Validation}

Reviewer: While the text provides detailed information about the methodology, it could benefit from providing a bit more context about the significance and real-world implications of the study. Explain why understanding travel behavior and infrastructure planning is important for urban areas like Leon County.

Reviewer 2: - [x] @jvtcl to address this.

\input tables/demographic_validation.tex

\input figures/local_demographic_validation.tex

> Each of the three P-MEDM runs (one for each PUMA in the Tallahassee CBSA) resulted in a synthetic reconstruction of the ACS 90\% MOEs that conformed within 99\% of the published MOEs from the ACS SF (Table~\ref{tab:demographic-validation}. Mapped to Leon County block groups (Figure~\ref{fig:local-demographic-validation}), MOE Fit Rates were somewhat lower but still conform within 90\% of the ACS SF MOEs. In particular, we observed diminished performance in areas with large group quarters populations (e.g., college dormitories, prisons), as well as more sparsely populated rural and peri-urban portions of Leon County. 

Reviewer: - It was difficult to grasp the main point here. My point here is break them down into shorter, more digestible sentences. For example, instead of "Each of the three P-MEDM runs (one for each PUMA in the Tallahassee CBSA) resulted in a synthetic reconstruction of the ACS 90% MOEs that conformed within 99% of the published MOEs from the ACS SF," you could say "Each of the three P-MEDM runs (one for each PUMA in the Tallahassee CBSA) resulted in a synthetic reconstruction of the ACS 90% MOEs. These reconstructions conformed within 99% of the published MOEs from the ACS SF." Let me know how that sounds?

Reviewer 2: @kalyan678 We agree with this suggestion. 

- [x] @jvtcl to address this.

> Each of the three P-MEDM runs (one for each PUMA in the Tallahassee CBSA) resulted in a synthetic reconstruction of the ACS 90\% MOEs that conformed within 99\% of the published MOEs from the ACS SF (Table~\ref{tab:demographic-validation}. Mapped to Leon County block groups (Figure~\ref{fig:local-demographic-validation}), MOE Fit Rates were somewhat lower but still conform within 90\% of the ACS SF MOEs. In particular, we observed diminished performance in areas with large group quarters populations (e.g., college dormitories, prisons), as well as more sparsely populated rural and peri-urban portions of Leon County. 

Reviewer 3: somewhat -> use more formal words like 'relatively' or 'moderately'

Reviewer 2: - [x] @jvtcl to address this.

> Each of the three P-MEDM runs (one for each PUMA in the Tallahassee CBSA) resulted in a synthetic reconstruction of the ACS 90\% MOEs that conformed within 99\% of the published MOEs from the ACS SF (Table~\ref{tab:demographic-validation}. Mapped to Leon County block groups (Figure~\ref{fig:local-demographic-validation}), MOE Fit Rates were somewhat lower but still conform within 90\% of the ACS SF MOEs. In particular, we observed diminished performance in areas with large group quarters populations (e.g., college dormitories, prisons), as well as more sparsely populated rural and peri-urban portions of Leon County. 

Reviewer 3: Is there determined reasoning for why large population centers exhibit diminished performance? Can you mention that after concluding this sentence?

Reviewer 2: - [x] @jvtcl to address this.

Reviewer 2: > * @jvtcl to address this.

We feel that we don't have enough evidence from this single case study to provide a well-defined reason for why this occurs. A few potential reasons (population density, diversity of travel modes) are now provided in the text. We could explore this further in future work by repeating the case study analysis elsewhere.

\subsubsection{Mobility Validation}

\input figures/local_win_cca.tex

> Our mobility simulation more realistically recreated between-destination demographics ($R^2$ = 0.74) than within-destination demographics ($R^2$ = 0.41) (Table~\ref{tab:demographic-validation}). Inspecting the local within-destination scores, mapped in Figure~\ref{fig:local-win-cca}, we observed generally greater correspondence between synthetic and observed POI demographics near Tallahassee's downtown core, Florida State University, and Florida A\&M University, with diminished performance in suburban and outlying areas of Leon County. 

Reviewer: - It would be really helpful if you can provide more details about the limitations and challenges faced during the mobility validation exercise and the modest success achieved. This will help readers understand the context and potential areas for improvement.

Reviewer 2: - [x] @jvtcl to address this.

\subsection{Limitations}\label{section:limitations}

> We found that overall, Likeness approximates travel to non-anchor (grocery store) POIs modestly well. However, contrasted with anchor activities (work/school) demonstrated in \cite{likeness-scipy-paper-2022}, non-anchor activities, like grocery store visits, occur at locations that do not have defined capacities (e.g., employee or enrollment sizes) are a greater challenge to represent realistically. Multiple factors related to data inputs and model assumptions may have confounded our results:

Reviewer 3: confounded -> more formal 'negatively impacted' or 'negatively influenced'

Reviewer 2: - [x] @jvtcl to address this.

Reviewer 2: @ameyxd We are in favor of leaving the wording here as-is, since we are identifying [confounding factors](https://www.sciencedirect.com/topics/nursing-and-health-professions/confounding-variable) associated with the inputs to our integrated model.

Reviewer 2: Ended up revising this sentence to: "Multiple confounding factors related to data inputs and model assumptions may have affected our results..."

\begin{itemize}

>     \item The clearest \textbf{data-specific} challenge is temporal mismatch between the synthetic population (2019) and POI visit data (2023). We have yet to increment our ACS year due to sporadic issues of non-conforming geography between the 2010 PUMAs and 2020 block groups/tracts that we hope to resolve starting with the forthcoming 2022 ACS releases, which will begin to use 2020 PUMAs\footnote{https://www.census.gov/programs-surveys/acs/news/data-releases/2022/release.html}. We were also limited to only one month of POI visit data, but we hope to expand its scope for future validation exercises.

Reviewer 3: Reason for emboldening 'data-specific' and 'model-specific'?

Reviewer 2: @ameyxd We emboldened this text to highlight the subject of each item. We could also place this text before the descriptions if you think that would be clearer.

Reviewer 2: - [x] @jvtcl to address this.

    \item Several large \textbf{model-specific} assumptions were made that could confound our results, particularly that 1) agents simultaneously travel to grocery stores, 2) agents only select one grocery store to visit, using only one mode of travel, and 3) travel to grocery stores only occurs between home and work. These assumptions can be updated by incorporating information from time-use and travel surveys into Likeness, thereby better reflecting the times of day that different demographic cohorts access various activities \cite{macal2018chisim}.

>     \item Going forward, we also hope to tighten our assumptions about the feasibility of POI access relative to the various travel modes. For example, in our current assignment process (Figure~\ref{fig:mode_assign_workflow}) agents who we could not label with a defined travel mode were considered walkers as a fallback for unavailability of a bus route in their block group of residence. Future iterations will refine this decision process by considering a distance threshold for agents to be labeled as either walkers or transit users (e.g., walking agents must be located within a reasonable distance of the closest POI, while agents that use bus service should reside near a bus stop, not merely be located in a block group that intersects a bus route). To ensure all agents have at least one feasible POI destination to access, we also plan to incorporate a greater variety of curated locations from ORNL's PlanetSense database \cite{thakur2015planetsense}. 

Reviewer 3: I do not see the purpose of having point 3 in this list, since it talks about potential changes you hope to make in future iterations. Consider having this as a conclusion of this section instead of a point.

Reviewer 2: @ameyxd We agree with this suggestion and will make the recommended changes. 

- [ ] ~~@jvtcl~~ @jGaboardi to address this.

Reviewer 2: To address @ameyxd's comment below, I added a few research directions we could investigate, one of them being "behavioral factors (willingness to travel given cost and impedances)". 

I moved the paragraph to the conclusion as recommended, but it feels redundant with this point. @jGaboardi, would you mind looking this over?

Author: We have reworded and the adjusted text is now up in this PR.

* @jvtcl Please mark complete

\end{itemize}

\subsection{Conclusion and Outlook}\label{section:conclusion-outlook}

> This paper presented enhancements and scaling for the Likeness spatial microsimulation toolkit -- including implementing batched population synthesis runs and large-scale transportation network generation -- to enable activity modeling with customizable agent populations based on real-world housing stock and transportation infrastructure for any metropolitan statistical area (MSA) in the United States. We demonstrated these capabilities by developing a mobility validation exercise for Leon County, FL. While the results of this initial exercise were only modestly successful, they provide new research questions that we plan to tackle as our workflow develops. 

Reviewer 3: The concluding statement is a bit vague. Please briefly point out one or two research questions that are evident from your initial exercise, and refrain from mentioning 'plan to tackle as our workflow develops' -> 'plan to answer in future work'.

Reviewer 2: - [x] @jvtcl to address this.

\subsection{Development Roadmap: 2023 - 2024} \label{section:dev-roadmap}

\begin{itemize}
    \item \textbf{Tooling for custom geographic extents.} The MSA-specific workflow demonstrated in this paper is limited in that it does not support custom geographic extents, limiting analysis, for example, of predominantly rural areas. 
    We are actively developing an approach to create residential synthetic populations for custom areas of interest (AOIs), supported by USA Structures, which will include national-scale coverage. 
    \item \textbf{Open-sourcing core packages.} Though we have yet to meet our goal of open-sourcing the suite of Likeness packages \cite{likeness-scipy-paper-2022}, we are still on track to release the core packages for residential population synthesis, \texttt{pymedm} and \texttt{livelike}, in 2023, with releases of \texttt{movelike} and \texttt{actlike} likely to follow in 2024.

>     \item \textbf{Packaging schema.} A further consideration related to open-sourcing is whether we should migrate from a confederated toolkit schema where each module is a semi-independent Python package, as is seen in the modern implementation of PySAL \cite{pysal_GA_2022}, to a single monolithic Python package with submodules. Each schema has benefits and this decision will require much consideration. 

Reviewer: May be, it's good to elaborate on the benefits and considerations associated with migrating from a confederated toolkit schema to a monolithic Python package. Discuss the implications and potential advantages of each approach.

Reviewer 2: - [x] @jGaboardi to address this.

Author: @jvtcl I have added some more explanatory language in for this. Please mark my portion as completed.

    \item \textbf{Consolidating visualization functionality.} We are in the process of consolidating functionality related to the visualization of input, processing, and results that have been used in an ad-hoc manner in the past. An initial push will be made for the inclusion of ``made-to-order'' population density hexbin plots and network-space allocation routes.

>     \item \textbf{Improving mobility modeling.} Realism in mobility modeling is a key area in which we are constantly striving for improvements, specifically transit. As stated previously in \nameref{section:likeness-expansion-scaling}, there is significant potential in exploring further integration of GTFS data for locally-accurate modeling.

Reviewer 3: Too subjective/informal, please rephrase the following: "Realism in mobility modeling is a key area in which we are constantly striving for improvements, specifically transit."

Reviewer 2: - [x] @jGaboardi  to address this.

Author: * I have revised that sentence.
* @jvtcl Please mark as completed.

>     \item \textbf{Optimization bottleneck.} As demonstrated in \nameref{section:results}, there is a clear hit in computational performance and runtime when solving  \texttt{actlike.ActivityAllocation} problems on increasingly larger model instances (e.g., more agents and more POIs). There are two paths to resolving this issue (which may be considered in concert): 1) Reviewing our modified Transportation Problem mixed-integer program (formulated in \cite{likeness-scipy-paper-2022}); and 2) Utilizing a new underlying solver engine, such as HiGHS\footnote{https://highs.dev/} \cite{highs_mpc_2018}.

Reviewer 3: Scalability issues are pretty common in simulation problems. I would like to see an additional sentence here, summarizing it and the general plan.

Reviewer 2: - [ ] @jvtcl to discuss tooling for generating synthetic populations with national coverage.
- [x] @jGaboardi to discuss challenges associated with scaling activity allocation and ways forward. Benefits of an engine like `HiGHS` towards this task?

Author: @jvtcl I have added some more explanatory language in for this. Please mark my portion as completed.

% This manuscript has been authored by UT-Battelle, LLC, under contract DE-AC05-00OR22725 with the US Department of Energy (DOE). The US government retains and the publisher, by accepting the article for publication, acknowledges that the US government retains a nonexclusive, paid-up, irrevocable, worldwide license to publish or reproduce the published form of this manuscript, or allow others to do so, for US government purposes. DOE will provide public access to these results of federally sponsored research in accordance with the DOE Public Access Plan (http://energy.gov/downloads/doe-public-access-plan).

% bibliography
\bibliographystyle{ieeetr}

> \bibliography{ mybib }

Reviewer 3: Overall, this is solid work, with quality content and descriptive diagrams/tables. It would benefit from the addition of some broader implications, insights, and details of the findings. Good job!

> \end{document}

Reviewer: Overall, the paper is well-written and informative. The quality of the writing is good. The content is well-structured and organized, making it easy to follow the flow of information. However, I feel some sentences and phrases could be simplified to enhance readability. Finally, the paper demonstrates a strong technical understanding and presents a detailed analysis of the research. It effectively communicates the objectives, methods, and results of the study.

Reviewer 2: @kalyan678 We appreciate your comments and will address readability and discussion points you mentioned in your review.
